File: aricode/sparring/challenges/07_mersenne/mersenne.go
// Go - Challenge 07: Mersenne Prime Verification
// Verify M31 = 2^31 - 1 = 2,147,483,647 is prime
// Recursive trial division against ~23,170 odd divisors

package main

import "os"

func isPrimeRec(n, d int) int {
	if n/d < d {
		return 1
	}
	if n%d == 0 {
		return 0
	}
	return isPrimeRec(n, d+2)
}

func main() {
	os.Exit(isPrimeRec(2147483647, 3))
}

File: aricode/sparring/challenges/07_mersenne/mersenne.py
#!/usr/bin/env python3
# Python - Challenge 07: Mersenne Prime Verification
# Verify M31 = 2^31 - 1 = 2,147,483,647 is prime
# Iterative trial division (Python uses bigints natively)

import sys

def is_prime(n):
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    d = 3
    while d * d <= n:
        if n % d == 0:
            return False
        d += 2
    return True

sys.exit(1 if is_prime(2147483647) else 0)
